pbr: T6430: Allow forwarding into VRFs by name as well as route table IDs

talmakion opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
My [previous PR](https://github.com/vyos/vyos-1x/pull/3581) for simply extending the table ID range was rejected, so I've prepared another PR closer to my own use case for this - "policy based route leaking" between VRFs. It may or may not line up with Bernhard's intention from the original ticket. 

This PR extends PBR syntax to allow "set vrf" as well as "set table", resolving table ID from an active VRF and then treating it the same as a "set table". 

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
* https://vyos.dev/T6430

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
 
## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
* pbr

## Proposed changes
<!--- Describe your changes in detail -->
* PBR can only target table IDs up to 200 and the previous PR to extend the range was rejected
* PBR with this PR can now also target VRFs directly by name, working around targeting problems for VRF table IDs outside the overlapping 100-200 range
* Validation ensures rules can't target both a table ID and a VRF name (internally they are handled the same)
* Added a simple accessor (get_vrf_table_id) for runtime mapping a VRF name to table ID, based on vyos.ifconfig.interface._set_vrf_ct_zone(). It does not replace that usage, as it deliberately does not handle non-VRF interface lookups (would fail with a KeyError). 
* Added smoketest to validate PBR 'set vrf' syntax

The use cases of this feature are limited and you will need to know *exactly* what you are doing to have it work properly, but for example, when I was experimenting with cross-VRF DNS lookup solutions, this would've been better than manipulating ip rules and nftables manually and externally. 

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

Rules apply OK:
```
set name MGT table '666'
set policy route TEST rule 100 destination address '8.8.8.8'
set policy route TEST rule 100 destination port '53'
set policy route TEST rule 100 protocol 'tcp_udp'
set policy route TEST rule 100 set vrf 'MGT'
set policy route TEST2 rule 100 destination address '10.11.12.77'
set policy route TEST2 rule 100 protocol 'tcp_udp'
set policy route TEST2 rule 100 set vrf 'default'
set policy route TEST2 rule 100 source port '53'
```

Looking at the results:
``` 
# sudo ip rule show
220:    from all lookup 220
254:    from all fwmark 0x7fffff01 lookup main
666:    from all fwmark 0x7ffffd65 lookup MGT 
1000:   from all lookup [l3mdev-table]
2000:   from all lookup [l3mdev-table] unreachable
32765:  from all lookup local
32766:  from all lookup main
32767:  from all lookup default
# sudo nft list ruleset
[...]
table ip vyos_mangle {
        chain VYOS_PBR_PREROUTING {
                type filter hook prerouting priority mangle; policy accept;
        }

        chain VYOS_PBR_POSTROUTING {
                type filter hook postrouting priority mangle; policy accept;
        }

        chain VYOS_PBR_UD_TEST {
                meta l4proto { tcp, udp } ip daddr 8.8.8.8 th dport 53 counter packets 0 bytes 0 meta mark set 0x7ffffd65 return comment "ipv4-route-TEST-100"
        }

        chain VYOS_PBR_UD_TEST2 {
                meta l4proto { tcp, udp } ip daddr 10.11.12.77 th sport 53 counter packets 0 bytes 0 meta mark set 0x7fffff01 return comment "ipv4-route-TEST2-100"
        }
}
[...]
```

Attaching the PBRs to interfaces and watching traffic flow with tcpdump works as expected, the same as `set table`. 

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
$ python3 /usr/libexec/vyos/tests/smoke/cli/test_policy_route.py 
test_pbr_group (__main__.TestPolicyRoute.test_pbr_group) ... ok
test_pbr_mark (__main__.TestPolicyRoute.test_pbr_mark) ... ok
test_pbr_mark_connection (__main__.TestPolicyRoute.test_pbr_mark_connection) ... ok
test_pbr_matching_criteria (__main__.TestPolicyRoute.test_pbr_matching_criteria) ... ok
test_pbr_table (__main__.TestPolicyRoute.test_pbr_table) ... ok
test_pbr_vrf (__main__.TestPolicyRoute.test_pbr_vrf) ... ok

----------------------------------------------------------------------
Ran 6 tests in 32.673s

OK
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

talmakion commented:
I want to update this to also implement this for policy local-route, so withdrawing to draft.

sever-sever commented:
I guess we should not change two components even if they are similar in one PR
There should be another PR for the `policy local-route` as it does not use nftables at all.

talmakion commented:
Fair call - I'll put together another PR for local-route tomorrow. This one should be already good to go if it passes review.

talmakion commented:
Updated with smoke tests as requested by @nicolas-fort

talmakion commented:
Noticed a minor issue with completion ordering & missing extras while working on the local-route version of this, fixed.

c-po commented on src/conf_mode/policy_route.py:
```diff
@@ -152,15 +156,26 @@ def apply_table_marks(policy):
+                        set_vrf = dict_search_args(rule_conf, 'set', 'vrf')
+                        if set_vrf:
+                            if set_vrf == 'default':
+                                vrf_table_id = '254'
```
'254' is a recurring number, please introduce a global variable referencing this number so we have a single point of contact. Other then that, LGTM

c-po commented on python/vyos/firewall.py:
```diff
@@ -473,11 +474,20 @@ def parse_rule(rule_conf, hook, fw_name, rule_id, ip_name):
+            vrf_name = rule_conf['set']['vrf']
+            # NOTE: VRF->table ID lookup depends on the VRF iface already existing. 
+            if vrf_name == 'default':
+                table = '254'
```
'254' is a recurring number, please introduce a global variable referencing this number so we have a single point of contact. Other then that, LGTM


Can we import the magic value of 254 referenced in both firewall.py and policy_route.py from a single file? As it's policy based firewall.py should import the variable from policy_route.py

talmakion replied:
Do you really want something in vyos.* to import from conf_scripts? 

There's an equivalent PR for policy local-route coming when I write the smoketests. Stashing the reference somewhere central does make sense. The standard tables & their IDs are used in a few spots by VRF configuration as well. 

Would vyos.defaults or vyos.utils.network be more appropriate? Maybe something new and more specific?

c-po replied:
> Do you really want something in vyos.* to import from conf_scripts?

No, thanks for rethinking!

Please move it to `vyos.defaults`

talmakion replied:
Is something like this too verbose?
```
rt_symbolic_names = {
  # Standard tables for Linux & reserved IDs for VyOS
  'default': 253, # Confusingly, a final fallthru, not the default. 
  'main': 254,    # The actual global table used by iproute2 unless told otherwise. 
  'local': 255,   # Special kernel loopback table
}

rt_global_vrf = rt_symbolic_names['main']
rt_global_table = rt_symbolic_names['main']
```
If there's other special tables reserved by VyOS in future, they can be centrally recorded in vyos.defaults.rt_symbolic_names.

c-po replied:
No its properly verbose - everyone knows what's what. Thanks

talmakion replied:
Requested changes pushed.

Commit:
pbr: T6430: Allow forwarding into VRFs by name as well as route table IDs

* PBR can only target table IDs up to 200 and the previous PR to extend the
  range was rejected
* PBR with this PR can now also target VRFs directly by name, working around
  targeting problems for VRF table IDs outside the overlapping 100-200 range
* Validation ensures rules can't target both a table ID and a VRF name
  (internally they are handled the same)
* Added a simple accessor (get_vrf_table_id) for runtime mapping a VRF name
  to table ID, based on vyos.ifconfig.interface._set_vrf_ct_zone().
  It does not replace that usage, as it deliberately does not handle non-VRF
  interface lookups (would fail with a KeyError).
* Added route table ID lookup dict, global route table and VRF table defs
  to vyos.defaults. Table ID references have been updated in code touched
  by this PR.
* Added a simple smoketest to validate 'set vrf' usage in PBR rules

talmakion commented:
Updated as requested with central default table IDs, manual tests and smoketests all OK. 

Additionally, in the spots where the table IDs were being defined as strings, they're now integers. Casts tweaked to match. They were always cast to ints anyway for the fwmark calculation. Multiple and overlapping table rules still work in tests.

c-po committed:
pbr: T6430: refactor to use vyos.utils.network.get_vrf_tableid()

Commit 452068ce78 ("interfaces: T6592: moving an interface between VRF instances
failed") added a similar but more detailed implementation of get_vrf_table_id()
that was added in commit adeac78ed of this PR. Move to the common available
implementation.

c-po commented:
I did a last minor change in 9b99a01653e3315b1abc9ef98824ca71bd283047 which dropped your implementation of get_vrf_table_id, as we have this as a common utility function as of last week.
